Compiling collections: a lot of cyclic reference errors

While trying to compile standard scala collections cyclic reference errors are very common:

```scala
GenericTraversableTemplate.scala:30: error: cyclic reference involving trait GenTraversable
trait GenericTraversableTemplate[+A, +CC[X] <: GenTraversable[X]] extends HasNewBuilder[A, CC[A] @uncheckedVariance] {

```

```scala
GenericSetTemplate.scala:16: error: cyclic reference involving trait GenSet
trait GenericSetTemplate[A, +CC[X] <: GenSet[X]] extends GenericTraversableTemplate[A, CC] {
```

```scala
GenericParTemplate.scala:28: error: cyclic reference involving trait ParIterable
trait GenericParTemplate[+A, +CC[X] <: ParIterable[X]]
```

--- a/src/dotty/tools/dotc/core/TypeApplications.scala
+++ b/src/dotty/tools/dotc/core/TypeApplications.scala
@@ -11,6 +11,7 @@ import Names._
 import NameOps._
 import Flags._
 import StdNames.tpnme
+import typer.Mode
 import util.Positions.Position
 import config.Printers._
 import collection.mutable
@@ -144,10 +145,23 @@ class TypeApplications(val self: Type) extends AnyVal {
           println(s"precomplete decls = ${self.typeSymbol.unforcedDecls.toList.map(_.denot).mkString("\n  ")}")
         }
         val tparam = tparams.head
-        val arg1 =
-          if ((tparam is HigherKinded) && !arg.isLambda && arg.typeParams.nonEmpty)
-            arg.EtaExpand
-          else arg
+        def needsEtaExpand =
+          try {
+            (tparam is HigherKinded) && !arg.isLambda && arg.typeParams.nonEmpty
+          }
+          catch {
+            case ex: CyclicReference =>
+              if (ctx.mode.is(Mode.Scala2Unpickling))
+                // When unpickling Scala2, we might run into cyclic references when
+                // checking whether eta expansion is needed or eta expanding.
+                // (e.g. try compile collection/generic/GenericTraversableTemplate.scala).
+                // In that case, back out gracefully. Ideally, we should not have
+                // underdefined pickling data that requires post-transformations like
+                // eta expansion, but we can't change Scala2's.
+                false
+              else throw ex
+          }
+        val arg1 = if (needsEtaExpand) arg.EtaExpand else arg
         val tp1 = RefinedType(tp, tparam.name, arg1.toBounds(tparam))
         matchParams(tp1, tparams.tail, args1)
       case nil => tp

--- a/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala
+++ b/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala
@@ -12,6 +12,7 @@ import scala.collection.{ mutable, immutable }
 import scala.collection.mutable.{ ListBuffer, ArrayBuffer }
 import scala.annotation.switch
 import typer.Checking.checkNonCyclic
+import typer.Mode
 import io.AbstractFile
 import scala.util.control.NonFatal
 
@@ -673,7 +674,7 @@ class ClassfileParser(
 
       def unpickleScala(bytes: Array[Byte]): Some[Embedded] = {
         val unpickler = new unpickleScala2.Scala2Unpickler(bytes, classRoot, moduleRoot)(ctx)
-        unpickler.run()
+        unpickler.run()(ctx.addMode(Mode.Scala2Unpickling))
         Some(unpickler)
       }
 

--- a/src/dotty/tools/dotc/core/unpickleScala2/Scala2Unpickler.scala
+++ b/src/dotty/tools/dotc/core/unpickleScala2/Scala2Unpickler.scala
@@ -17,6 +17,7 @@ import printing.Printer
 import io.AbstractFile
 import util.common._
 import typer.Checking.checkNonCyclic
+import typer.Mode
 import PickleBuffer._
 import scala.reflect.internal.pickling.PickleFormat._
 import Decorators._
@@ -520,7 +521,7 @@ class Scala2Unpickler(bytes: Array[Byte], classRoot: ClassDenotation, moduleClas
   class LocalUnpickler extends LazyType {
     def startCoord(denot: SymDenotation): Coord = denot.symbol.coord
     def complete(denot: SymDenotation)(implicit ctx: Context): Unit = try {
-      def parseToCompletion(denot: SymDenotation) = {
+      def parseToCompletion(denot: SymDenotation)(implicit ctx: Context) = {
         val tag = readByte()
         val end = readNat() + readIndex
         def atEnd = readIndex == end
@@ -566,7 +567,8 @@ class Scala2Unpickler(bytes: Array[Byte], classRoot: ClassDenotation, moduleClas
         }
         // println(s"unpickled ${denot.debugString}, info = ${denot.info}") !!! DEBUG
       }
-      atReadPos(startCoord(denot).toIndex, () => parseToCompletion(denot))
+      atReadPos(startCoord(denot).toIndex,
+          () => parseToCompletion(denot)(ctx.addMode(Mode.Scala2Unpickling)))
     } catch {
       case ex: RuntimeException => handleRuntimeException(ex)
     }

--- a/src/dotty/tools/dotc/typer/Mode.scala
+++ b/src/dotty/tools/dotc/typer/Mode.scala
@@ -79,5 +79,8 @@ object Mode {
    */
   val ImplicitExploration = newMode(12, "ImplicitExploration")
 
+  /** We are currently unpickling Scala2 info */
+  val Scala2Unpickling = newMode(13, "Scala2Unpickling")
+
   val PatternOrType = Pattern | Type
 }

